File: UnacademyFreeLiveClass/HashMap.py
class Node:
	key = ""
	value = 0
	next = None
	def __init__(self, key, value):
		self.key = key
		self.value = value

class HashMap:
	cs = 0
	ts = 1
	arr = []

	def __init__(self, ts):
		self.ts = ts
		self.arr = [None]*ts

	def hashfunc(self, key):
		sum = 0
		factor = 1
		for i in range(len(key)):
			sum = ((sum%self.ts) + (ord(key[i])*factor)%self.ts)%self.ts
			factor = ((factor%self.ts) * (37%self.ts))%self.ts;

		return sum

	def rehash(self):
		print("Now rehashing")
		old_size = self.ts
		self.ts *= 2
		li = self.arr
		self.arr = [None]*self.ts
		self.cs = 0
		for i in range(old_size):
			temp = li[i]
			while temp != None:
				self.insert(temp.key, temp.value)
				temp = temp.next

	def insert(self, key, value):
		bi = self.hashfunc(key)
		node = Node(key, value)
		if self.arr[bi] == None:
			self.arr[bi] = node 
		else:
			# collision
			node.next = self.arr[bi]
			self.arr[bi] = node 
		self.cs += 1
		load_factor = self.cs / self.ts
		print(load_factor, self.cs, self.ts)
		if load_factor > 0.5:
			self.rehash()

	def search(self, key):
		bi = self.hashfunc(key)
		temp = self.arr[bi]
		while temp != None:
			if(temp.key == key):
				return temp.value
			temp = temp.next
		return None

	def update(self, key, value):
		if(self.search(key) == None):
			return False
		bi = self.hashfunc(key)
		temp = self.arr[bi]
		while(temp != None):
			if temp.key == key:
				temp.value = value
			temp = temp.next
		return True

	def display(self):
		for i in range(self.ts):
			temp = self.arr[i]
			if(temp == None):
				print("empty")
			else:
				while temp != None:
					print(f"key: {temp.key} -> value: {temp.value}", end = " | ")
					temp = temp.next
				print()


hashmap = HashMap(2)
hashmap.insert("mango", 1)
hashmap.insert("apple", 2)
hashmap.insert("appy", 11)
hashmap.insert("apipie", 81)
hashmap.insert("banana", 19)
hashmap.insert("bananpie", 100)
hashmap.insert("mangopie", 100)
hashmap.insert("applepie", 100)
hashmap.insert("api", 100)
hashmap.display()

File: UnacademyCourse/IntroToDS_CPIT_SEP/Lecture8/hashmap.cpp
#include<iostream>
#define ll long long int
template<typename T>
struct Node {
    std::string key;
    T value;
    Node<T>* next;

    Node(std::string k, T v) {
        key = k;
        value = v;
        next = NULL;
    }
};

template<typename T>
struct HashMap {
    int cs;
    int ts;
    Node<T> **arr;
    HashMap(int total_size = 2) {
        ts = total_size;
        cs = 0;
        arr = new Node<T>*[ts];
        for(int i = 0; i < ts; i++) {
            arr[i] = NULL;
        }
    }

    void display() {
        for(int i = 0 ; i < ts; i++) {
            Node<T> *temp = arr[i];
            if(temp == NULL) {
                std::cout<<"empty\n";
                continue;
            } else {
                while(temp != NULL) {
                    std::cout<<"("<<temp->key<<","<<temp->value<<")->";
                    temp = temp->next;
                }
                std::cout<<"\n";
            }
        }
    }
    int hashFunc(std::string key) {
        ll sum = 0;
        ll factor = 1;
        for(int i = 0 ; i < key.size(); i++) {
            sum = ((sum%ts) + ((int)key[i]*factor)%ts)%ts;
            factor = (factor%ts * 37%ts)%ts;
        }
        return (int)sum;
    }

    void rehash() {
        int old_size = ts;
        ts *= 2;
        Node<T>** oldTable = arr;
        arr = new Node<T>*[ts];
        for(int i = 0; i < ts; i++) {
            arr[i] = NULL;
        }
        for(int i = 0; i < old_size; i++) {
            Node<T> *temp = oldTable[i];
            while(temp != NULL) {
                insert(temp->key, temp->value);
                temp = temp->next;
            }
        }
    }

    void insert(std::string key, T value) {
        int bi = hashFunc(key);
        Node<T> * new_node = new Node<T>(key, value);
        if(search(key) != NULL) {
            Node<T> *temp = arr[bi];
            while(temp != NULL) {
                if(temp->key == key) {
                    temp->value = value;
                    return;
                }
                temp = temp->next;
            }
        }
        
        if(arr[bi] == NULL) {
            arr[bi] = new_node;
        } else {
            new_node->next = arr[bi];
            arr[bi] = new_node;
        }
        cs++;
        double loadFactor = (double)(cs) / ts;
        if(loadFactor > 0.5) {
            rehash();
        }
    }

    T search(std::string key) {
        int bi = hashFunc(key);
        Node<T> *temp = arr[bi];
        while(temp != NULL) {
            if(temp->key == key) {
                return temp->value;
            }
            temp = temp->next;
        }
        return NULL;
    }

};

int main(int argc, char const *argv[])
{
    /* code */
    HashMap<int> hash;
    hash.insert("mango", 27);
    hash.insert("apple", 7);
    hash.insert("mango", 2);
    hash.insert("grapes", 217);
    hash.insert("mango", 127);
    hash.insert("grapes", 1127);
    hash.insert("apple", 207);
    hash.insert("applepie", 227);
    hash.insert("pie", 7);
    hash.insert("mango", 74);
    hash.display();
    std::cout<<hash.search("pie");
    return 0;   
}
